packages/slate-react/src/components/editor.tsx

```diff
@@ -0,0 +1,488 @@
+        const { inputType } = event
+
+        // These two types occur while a user is composing text and can't be
+        // cancelled. Let them through and wait for the composition to end.
+        if (
+          inputType === 'insertCompositionText' ||
+          inputType === 'deleteCompositionText'
+        ) {
+          return
+        }
+
+        const [targetRange] = event.getTargetRanges()
+
+        if (targetRange && inputType !== 'deleteContent') {
+          const range = editor.toSlateRange(targetRange)
+          editor.select(range)
+        }
+
+        event.preventDefault()
+        editor.onBeforeInput(event)
+      }
+    },
+    []
+  )
+
+  // On native `selectionchange` event, trigger the `onSelect` handler. This is
+  // needed to account for React's `onSelect` being non-standard and not firing
+  // until after a selection has been released. This causes issues in situations
+  // where another change happens while a selection is being dragged.
+  const onDOMSelectionChange = useCallback(
+    debounce(() => {
+      console.log('selectionchange')
+      if (!readOnly && !state.isComposing && !state.isUpdatingSelection) {
+        const el = editor.toDomNode(value)
+        const { activeElement } = window.document
+
+        if (activeElement === el) {
+          const domSelection = window.getSelection()
+          state.latestElement = activeElement
+
+          if (domSelection) {
+            const range = editor.toSlateRange(domSelection)
+            console.log('select', domSelection, range)
+            editor.select(range)
+            IS_FOCUSED.set(editor, true)
+          }
+        } else {
+          IS_FOCUSED.delete(editor)
+        }
+      }
+    }, 100),
+    []
+  )
+
+  return (
+    <EditorContext.Provider value={editor}>
+      <ReadOnlyContext.Provider value={readOnly}>
+        <FocusedContext.Provider value={editor.isFocused()}>
+          <div
+            // COMPAT: The Grammarly Chrome extension works by changing the DOM
+            // out from under `contenteditable` elements, which leads to weird
+            // behaviors so we have to disable it like editor. (2017/04/24)
+            data-gramm={false}
+            role={readOnly ? undefined : props.role || 'textbox'}
+            {...attributes}
+            data-slate-editor
+            data-slate-node="value"
+            contentEditable={readOnly ? undefined : true}
+            suppressContentEditableWarning
+            ref={ref}
+            style={{
+              // Prevent the default outline styles.
+              outline: 'none',
+              // Preserve adjacent whitespace and new lines.
+              whiteSpace: 'pre-wrap',
+              // Allow words to break if they are too long.
+              wordWrap: 'break-word',
+              // Allow for passed-in styles to override anything.
+              ...style,
+            }}
+            onBlur={useCallback((event: React.FocusEvent) => {
```

Comment:
How would you attach an `onFocus` and `onBlur` hook now?

Reply:
I was thinking, but haven’t gotten around to implementing, that maybe we could call any extra `on*` handlers passed in that clash first. And then maybe check to see if propagation is stopped and that would let you completely override the defaults if needed? Dunno.

Reply:
That sounds good, I think it's important to be be able to stop the default behavior.